### ISSUE TITLE
REPL ignores jobName DataFlow parameter

### ISSUE THREAD

**Issue by arturroo** [spotify/scio#2019]
When I run REPL on DataFlow it sets jobName based on my username and not on jobName parameter for DataFlow. My Google IAM username consists '_' underscore sign and those signs are not allows in DataFlow job names, so REPL starts, but any action that needs DF jobs breaks with error: 
```
  at org.apache.beam.sdk.util.InstanceBuilder.buildFromMethod(InstanceBuilder.java:214)
  ... 62 more
Caused by: java.lang.IllegalArgumentException: JobName invalid; the name must consist of only the characters [-a-z0-9], starting with a letter and ending with a letter or number
  at org.apache.beam.vendor.guava.v20_0.com.google.common.base.Preconditions.checkArgument(Preconditions.java:122)
```
Workaroud is to set jobName manually when REPL is running:
```
scio> sc.setJobName("scio-repl-af")
scio> sc.options.getJobName
res2: String = scio-repl-af
scio> val mytable = sc.bigQueryTable("precog:streaming.mytab")
...
mytable: com.spotify.scio.values.SCollection[com.spotify.scio.bigquery.TableRow] = com.spotify.scio.values.SCollectionImpl@1263664f
scio>
```

Please allow overwriting Scio jobName with specified --jobName parameter to DataFlow.

**Comment by arturroo** [comment]
I think this might be fixed by checking if jobName is set in options during Scio Context initialization here
https://github.com/spotify/scio/blob/master/scio-repl/src/main/scala/com/spotify/scio/repl/ReplScioContext.scala
```Scala
class ReplScioContext(options: PipelineOptions, artifacts: List[String])
    extends ScioContext(options, artifacts) {

  this.setAppName("sciorepl")
  this.setJobName(s"""sciorepl-${CoreSysProps.User.value}-${System.currentTimeMillis()}""")
```
So I will try to fix it.

**Comment by regadas** [comment]
Thanks for reporting this! Can you normalize the user as well?

**Comment by arturroo** [comment]
Hi, 

Yes, I would like to normalize the username at the beginning. This would remove need for workaround.
Next I would like to add additional parameter --jobNamePostfix.
So in https://github.com/spotify/scio/blob/master/scio-repl/src/main/scala/com/spotify/scio/repl/ReplScioContext.scala
I would change line 27
```
this.setJobName(s"""sciorepl-${CoreSysProps.User.value}-${System.currentTimeMillis()}""")
```
to
```

  /** Overwrite job name to default for REPL session */
  this.setJobName("sciorepl-%s-%d".format(
    CoreSysProps.User.value.replaceAll("[^a-z0-9]", "0"),
    System.currentTimeMillis()
  ))
```
I see that in another issue jobName was made stiff. I think for REPL it is a good idea to have part of jobName stiff and part custom. For this one have to implement custom PipelineOptions like here:
https://cloud.google.com/dataflow/docs/guides/specifying-exec-params

I tried to do it quickly but I do not know where to import trait with custom options.
I think it should be Scala trait, because it is similar to Java interface, right?

Do you think it is OK?

**Comment by regadas** [comment]
Hi @arturroo to be honest I think i would prefer to override the job name through standard `--jobName` instead of introducing a new one.

```scala
  if (options.getJobName == null) {
    val normalizedUser = CoreSysProps.User.value.replaceAll("[^a-z0-9]", "0")
    this.setJobName(s"sciorepl-$normalizedUser-${System.currentTimeMillis()}")
  }
```

**Comment by arturroo** [comment]
Hi @regadas this would be great, but options are already parsed and set in class ScioILoop line 52
```
  try {
    PipelineOptionsFactory.fromArgs(args: _*).create()
  }
```
So it will never be null because in Apache Beam class JobNameFactory sets default value of JobName so in options passed to Scio Repl it is always set to Beam default or users custom value. 
This is Beam's default jobName from JobNameFactory class:
```
return String.format(
          "%s-%s-%s-%s", normalizedAppName, normalizedUserName, datePart, randomPart);
```
And since options do not have any flag, that the value from getJobName is default or custom we can either
- create regexp of Beam's default jobName string and recognize if in options is default and if yes then overwrite with Scio's Repl default 
```s"sciorepl-$normalizedUser-${System.currentTimeMillis()}"```
if not then leave user's custom jobName.
- or we can do my proposed solution, so for REPL ignore jobName flag as it is now and just add custom postfix parameter that will be added to stiff ScioRepl jobName for user's flexibility.

What do you think?

**Comment by regadas** [comment]
Ahh yes forgot about that! nice catch! 

So, in this case I would actually not even set `jobName`. Taking advantage of the `jobName` default format (which is "almost" the same as ours) we just need to set the `appName` to `sciorepl`. This needs to be done when we actually parse the args though.

**Comment by arturroo** [comment]
Yeah, thx.

I thought that earlier issue that hat this stiff jobName made was because you wanted people to force static string in jobname to have.
The difference between this stiff jobName and default Beam is randomPart at the end, and no "sciorepl-" at the beginning. So if we agree that Repl Apps can have the same rights as regular data flow apps then yes, better to remove this line completely and leave default as Beam has it.

I think I have nothing more to say :) So what do you think?

**Comment by regadas** [comment]
to be honest i like the idea of being able to override the default appName and the jobName that we have since we allow passing beam options from CLI.

**Comment by arturroo** [comment]
OK, great. So I'll simply remove those 2 lines that sets the stiff values of appName and jobName.
I confirm on my testing env, that it works and Scio Repl properly recognizes both CLI parameters.
```
artur_fejklowicz@cloudshell:~/scio-repl (precog)$ cat ./run-snapshot.sh
#!/usr/bin/env bash
# afejklowicz 2019-07-08
# start Scio REPL from Snapshot
java -jar \
        -Dbigquery.project=precog \
        scio-repl-0.8.0-SNAPSHOTv2.jar \
        --project=precog \
        --tempLocation=gs://devzh/staging \
        --runner=DataflowRunner \
        --region=europe-west1 \
        --zone=europe-west6-a \
        --jobName=scio-repl-af--`date +'%Y-%m-%d-%H-%M-%S'` \
        --appName=scio-repl-aftest
artur_fejklowicz@cloudshell:~/scio-repl (precog)$ ./run-snapshot.sh
Welcome to
                 _____
    ________________(_)_____
    __  ___/  ___/_  /_  __ \
    _(__  )/ /__ _  / / /_/ /
    /____/ \___/ /_/  \____/   version 0.8.0-SNAPSHOT
Using Scala version 2.12.8 (OpenJDK 64-Bit Server VM, Java 1.8.0_212)
Type in expressions to have them evaluated.
Type :help for more information.
BigQuery client available as 'bq'
[main] WARN com.spotify.scio.VersionUtil$ - Using a SNAPSHOT version of Scio: 0.8.0-SNAPSHOT
Scio context available as 'sc'
**scio> sc.options.getJobName
res0: String = scio-repl-af--2019-07-08-22-53-52
scio> sc.optionsAs[org.apache.beam.sdk.options.ApplicationNameOptions].getAppName()
res1: String = scio-repl-aftest**
scio> 
```

**Comment by regadas** [comment]
Closed via #2028 #2046. Thanks @arturroo
